src/Service.php

```diff
@@ -41,91 +45,66 @@ class Service {
-                    $brick = $bricks->$getter();
-
-                    if($brick) {
-                        $brickFieldDefinitions = $brick->getDefinition()->getFieldDefinitions();
-                        foreach($brickFieldDefinitions as $bfd) {
-                            if($bfd instanceof PermissionManyToManyRelation) {
-                                $permissionObjects = array_merge($permissionObjects, $brick->{'get' . $bfd->getName()}());
-                            }
-                            if($bfd instanceof PermissionManyToOneRelation) {
-                                $href = $object->{'get' . $bfd->getName()}();
-                                if($href) {
-                                    $permissionObjects[] = $href;
-                                }
-                            }
-                            if($bfd instanceof PermissionResource) {
-                                $permissions[$bfd->getName()] = $brick->{'get' . $bfd->getName()}();
-                            }
-
-                            if($fd instanceof DynamicPermissionResource) {
-                                $permissions = array_merge($permissions, $brick->{'get' . $fd->getName()}());
-                            }
-                        }
-                    }
-                }
-            }
+        foreach ($fieldDefinitions as $fieldDefinition) {
+            [$fieldPermissions, $fieldPermissionObjects] = $this->getPermissionsByFieldDefinition($object, $fieldDefinition);
+            $permissions = array_merge($permissions, $fieldPermissions);
+            $permissionObjects = array_merge($permissionObjects, $fieldPermissionObjects);
         }
 
-        // get permission resources from linked objects and merge them with permissions of given object
-        // - when permission is set to allow / deny directly in object, this is always used
-        // - otherwise optimistic merging is used -> once one permission is allowed, it stays that way
+        // get permission resources from related objects and merge them with permissions of base object
         $mergedPermissions = $permissions;
-        foreach($permissionObjects as $permissionObject) {
+        foreach ($permissionObjects as $permissionObject) {
             $objectPermissions = $this->getPermissions($permissionObject, $visitedIds);
+            $mergedPermissions = $this->mergeNestedObjectPermissions($mergedPermissions, $permissions, $objectPermissions);
+        }
 
-            foreach($objectPermissions as $key => $value) {
-                if(($permissions[$key] == self::INHERIT || !array_key_exists($key, $permissions)) && $mergedPermissions[$key] != self::ALLOW) {
-                    $mergedPermissions[$key] = $value;
-                }
+        $permissionsEvent = new PermissionsEvent($mergedPermissions, $object, $this);
+        \Pimcore::getEventDispatcher()->dispatch(PermissionsEvent::POST_GET, $permissionsEvent);
```

Comment:
should we inject event dispatcher directly via DI and avoid static pimcore call?

Reply:
Good point. I'll change that after the refactor is through.

Reply:
Updated and tested

Reply:
thank you very much!